File: Top-Interview-Questions/Trees-Easy-Collection/108-Convert-Sorted-Array-to-Binary-Search-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// Time Complexity: O(n)
// Space Complexity: O(h)
// Runtime: 12 ms, faster than 97.65% 
// Memory Usage: 20.5 MB, less than 90.46%
class Solution {
public:
    
    TreeNode* DFS(vector<int> &nums, int left, int right){
        if(left > right)
            return NULL;
        int mid = (left+right)/2;
        TreeNode* node = new TreeNode();
        node->val = nums[mid];
        node->left = DFS(nums, left, mid-1);
        node->right = DFS(nums, mid+1, right);
        
        return node;
    }
    
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        int left = 0, right = nums.size()-1;
        
        return DFS(nums, left, right);
    }
};

File: Top-Interview-Questions/Trees-Easy-Collection/108-convert-sorted-array-to-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 * Runtime: 80 ms, faster than 34.63% of JavaScript online submissions for Convert Sorted Array to Binary Search Tree.
 * Memory Usage: 39.1 MB, less than 15.68% of JavaScript online submissions for Convert Sorted Array to Binary Search Tree.
 */
var sortedArrayToBST = function(nums) {
    if (!nums || nums.length === 0) {
        return null;
    }

    return buildTree(0, nums.length - 1, nums);
};

function buildTree (first, last, array) {
    if (first === last) {
        return new TreeNode(array[first]);
    } else if (first > last) {
        return null;
    }

    const mid = Math.floor((first + last) / 2);
    const node = new TreeNode(array[mid]);
    node.left = buildTree(first, mid - 1, array);
    node.right = buildTree(mid + 1, last, array);

    return node;
}